src/realm/sync/noinst/sync_schema_migration.cpp

```diff
@@ -0,0 +1,111 @@
+
+    Obj first = *table->begin();
+    REALM_ASSERT(first);
+    auto version = first.get<int64_t>(version_col);
+    auto time = first.get<Timestamp>(timestamp_col);
+    if (version != metadata_version) {
+        throw SyncSchemaMigrationFailed(
+            util::format("Unsupported sync schema migration metadata version: %1 vs %2, from %3", version,
+                         metadata_version, time));
+    }
+    return first.get<int64_t>(previous_schema_version_col);
+}
+
+void track_sync_schema_migration(TransactionRef wt, uint64_t previous_schema_version)
+{
+    REALM_ASSERT(wt);
+    TableRef table = wt->get_table(s_meta_schema_migration_table_name);
+    ColKey version_col, timestamp_col, previous_schema_version_col;
+    if (!table) {
+        table = wt->add_table_with_primary_key(s_meta_schema_migration_table_name, type_ObjectId, s_pk_col_name);
+        REALM_ASSERT(table);
+        version_col = table->add_column(type_Int, s_version_column_name);
+        timestamp_col = table->add_column(type_Timestamp, s_timestamp_col_name);
+        previous_schema_version_col = table->add_column(type_Int, s_previous_schema_version_col_name);
+    }
+    else {
+        version_col = table->get_column_key(s_version_column_name);
+        timestamp_col = table->get_column_key(s_timestamp_col_name);
+        previous_schema_version_col = table->get_column_key(s_previous_schema_version_col_name);
+    }
+    REALM_ASSERT(version_col);
+    REALM_ASSERT(timestamp_col);
+    REALM_ASSERT(previous_schema_version_col);
+
+    // A migration object may exist if the migration was started in a previous session.
+    if (table->is_empty()) {
+        table->create_object_with_primary_key(ObjectId::gen(),
+                                              {{version_col, metadata_version},
+                                               {timestamp_col, Timestamp(std::chrono::system_clock::now())},
+                                               {previous_schema_version_col, int64_t(previous_schema_version)}});
+    }
+    else {
+        auto first = *table->begin();
+        auto version = first.get<int64_t>(version_col);
+        auto time = first.get<Timestamp>(timestamp_col);
+        if (version != metadata_version) {
+            throw SyncSchemaMigrationFailed(
+                util::format("Unsupported sync schema migration metadata version: %1 vs %2, from %3", version,
+                             metadata_version, time));
+        }
+        uint64_t schema_version = first.get<int64_t>(previous_schema_version_col);
+        if (schema_version != previous_schema_version) {
+            throw SyncSchemaMigrationFailed(
+                util::format("Cannot continue sync schema migration with different previous schema version (existing "
+                             "previous_schema_version=%1, new previous_schema_version=%2)",
+                             schema_version, previous_schema_version));
+        }
```

Comment:
>Do we ever clear the value in the previous_schema_version_col (e.g. after the client bootstraps at the new schema version)?

we don't need to clear it out because the realm is deleted at that point (the bootstrap at the new schema version happens on the new realm where the metadata object does not exist anymore). So if a migration is successful, the object is gone.

>It seems like this would error out if we ever did a second schema migration

That's not the case. We allow migrations even if a previous migrations did not complete (check `Migrate to new schema version after migration to intermediate version is interrupted` test). The error here is to be safe (if anything it would catch the server doing something wrong).

>(e.g. client app is updated and metadata realm remains in place).

It may be confusing, but the metadata tables are actually created in the client's realm, not the medatada realm.

>Or do we need to retain this value for some other reason (e.g. always client reset at the previous schema version)?

This value is needed just for client resets. And the clients indeed client reset at previous schema version, otherwise the app would crash with a schema mismatch due to the breaking changes.

Reply:
Thanks for the explanation - since the read/write transaction is being passed in, I was wondering if this was the case. I saw that test, but wasn't sure if it was passing due to the failures.